hibernate-core/src/main/java/org/hibernate/loader/plan/exec/internal/AbstractLoadPlanBasedLoader.java

```diff
@@ -471,22 +472,17 @@ protected void advance(final ResultSet rs, final RowSelection selection) throws
 	private synchronized ResultSet wrapResultSetIfEnabled(final ResultSet rs, final SessionImplementor session) {
 		// synchronized to avoid multi-thread access issues; defined as method synch to avoid
 		// potential deadlock issues due to nature of code.
-		if ( session.getFactory().getSessionFactoryOptions().isWrapResultSetsEnabled() ) {
-			try {
-				if ( log.isDebugEnabled() ) {
-					log.debugf( "Wrapping result set [%s]", rs );
-				}
-				return session.getFactory()
-						.getServiceRegistry()
-						.getService( JdbcServices.class )
-						.getResultSetWrapper().wrap( rs, retreiveColumnNameToIndexCache( rs ) );
-			}
-			catch(SQLException e) {
-				log.unableToWrapResultSet( e );
-				return rs;
+		try {
+			if ( log.isDebugEnabled() ) {
+				log.debugf( "Wrapping result set [%s]", rs );
 			}
+			return session.getFactory()
```

Comment:
A great deal of these operations could be moved out of the synchronized block, even when wrapping is enabled.

I'll take your idea and go one step further, I hope you don't mind ;-)